Add a Fade Start property to ReflectionProbe

Calinou opened the pull request:
The default value of `0.5` makes reflections from ReflectionProbes more visible. This significantly reduces the notorious "sky is visible through solid objects in reflections" issue that people tend to notice when setting up reflection probes in Godot. If you set it to `1.0` in an indoor scene, the issue should be entirely prevented now, provided the probe's origin is not within solid one-sided geometry.

Please test this on more complex scenes to check if the proposed default value is actually a good idea. It may be a bit too aggressive; perhaps `0.2` would be a better fit.

To get the old behavior back, set Fade Start to `0.0`.

See https://github.com/godotengine/godot-proposals/issues/3013.

**Testing project:** [test_reflection_probes_mirage_palace.zip](https://github.com/godotengine/godot/files/12717917/test_reflection_probes_mirage_palace.zip)

## Preview

Before (Fade Start 0.0) | After (Fade Start 0.5, *default*)
-|-
![Screenshot_20230925_192235 webp](https://github.com/godotengine/godot/assets/180032/767ac19e-1f9c-4ad1-9d13-061f40148fbb) | ![Screenshot_20230925_192242 webp](https://github.com/godotengine/godot/assets/180032/562b4b91-d2db-4ae6-ab9c-33b5621a3399)

Fade Start 1.0 is meant to be used when you have a single probe cover an entire room, so it's not the best fit for this situation, but here it is anyway for comparison purposes:

After (Fade Start 0.85) | After (Fade Start 1.0)
-|-
![Screenshot_20230925_192250 webp](https://github.com/godotengine/godot/assets/180032/8ce1d7a0-f7b6-4593-96ee-34b21156152d) | ![Screenshot_20230925_192254 webp](https://github.com/godotengine/godot/assets/180032/5cf9398f-f3eb-4e25-ae31-6b3e268a95af)

## TODO

- [x] Implement in Compatibility.

Calinou committed:
Add a Fade Start property to ReflectionProbe

The default value of 0.5 makes reflections from ReflectionProbes
more visible. To get the old behavior back, set Fade Start to 0.0.

jcostello commented:
Works fine but with `Fade Start` at 0 between a inside and outside probe you get that fine line

![image](https://user-images.githubusercontent.com/1168582/174067344-a376cb3f-7f74-476f-ac17-b12a0955ae0b.png)

Calinou commented:
> Works fine but with `Fade Start` at 0 between a inside and outside probe you get that fine line

Fade Start = 0 is the existing behavior. Do you get this issue without this PR?

jcostello commented:
@Calinou sorry, I ment 1.0

Calinou commented:
> @Calinou sorry, I ment 1.0

In that case, I'm not sure how this can be prevented due to floating-point precision issues. You can make ReflectionProbes slightly overlap each other to work around this.

jcostello commented:
> 

They are overlap already. That happen at the end of the Reflection Probes set as `inside: false`

jcostello commented:
> > 
> 
> They are overlap already. That happen at the end of the Reflection Probes set as `inside: false`
> 
> Edit: One way of solving this is to add a parameter fade away the sky from the edge of the reflection probes and other parameter to fade away the probe into another probe. This will fix transitions like this also
> 
> ![image](https://user-images.githubusercontent.com/1168582/174392878-fed118a9-ea5c-412f-8bac-412fa9e292fa.png)

That way you can completely fade away the sky reflection in the edge but smootly fade away the probe with other probes and avoid this issue

![image](https://user-images.githubusercontent.com/1168582/174412472-baa33976-b69c-4553-a9ab-c16b81af6eb5.png)

clayjohn reviewed (commented):
I'm not sure that this addresses the use-case in https://github.com/godotengine/godot-proposals/issues/3013

This change may be helpful as it allows users to control where the blend starts and thus may help with some edge cases, but ultimately, the problem will still exist. Take for example, the example in the OP. With this change, the user could tweak the blend start so that very little light leaked into the room, by setting the fade start at roughly the room extents. The issue is this just pushes the blend portion outside of the room, so you will end up with occlusion artifacts outside of the room. What is needed is for us to save a depth map with the reflection probe and then use the depth map to blend.

Calinou commented:
One more situation where this popped up: https://www.reddit.com/r/godot/comments/10gux8y/why_is_the_skybox_overlaying_the_reflection/

In this particular example (mostly outdoor scene), increasing Fade Start would do the trick well.

> What is needed is for us to save a depth map with the reflection probe and then use the depth map to blend.

Is this cheap enough to be feasible on the Forward Mobile backend, let alone Compatibility?

atirut-w commented:
> What is needed is for us to save a depth map with the reflection probe and then use the depth map to blend.

Isn't that literally the same as adding lights with shadows to the scene? That can't be good for performance if it's enabled by default despite it fixing issues. I think being able to customize fade start is many times better.

AThousandShips commented on servers/rendering/renderer_rd/storage_rd/light_storage.cpp:
```diff
@@ -568,6 +575,13 @@ float LightStorage::reflection_probe_get_intensity(RID p_probe) const {
 
+float LightStorage::reflection_probe_get_fade_start(RID p_probe) const {
+	const ReflectionProbe *reflection_probe = reflection_probe_owner.get_or_null(p_probe);
+	ERR_FAIL_COND_V(!reflection_probe, 0);
```
```suggestion
	ERR_FAIL_NULL_V(reflection_probe, 0);
```

Calinou replied:
Note that LightStorage is using `ERR_FAIL_COND_V(!reflection_probe` all over the place already, so it may be better to fix that in a dedicated PR.

AThousandShips replied:
They will, but I'd like to not have to add these as well in my PR

AThousandShips replied:
Or rather my pr has been approved for this part so I'd have to do a separate one after in that case

AThousandShips replied:
And merged :)

AThousandShips commented on servers/rendering/renderer_rd/storage_rd/light_storage.cpp:
```diff
@@ -397,6 +397,13 @@ void LightStorage::reflection_probe_set_intensity(RID p_probe, float p_intensity
 
+void LightStorage::reflection_probe_set_fade_start(RID p_probe, float p_fade_start) {
+	ReflectionProbe *reflection_probe = reflection_probe_owner.get_or_null(p_probe);
+	ERR_FAIL_COND(!reflection_probe);
```
```suggestion
	ERR_FAIL_NULL(reflection_probe);
```

Calinou commented:
Rebased and tested again, it works as expected.

@jcostello I've fixed the issue you mentioned by decreasing the `MAX()` threshold in the `inv_fade_start` setter on the C++ side.

jcostello commented:
@Calinou nice, I'm testing this again

jcostello commented:
@Calinou The issue seems to be fixed. I love the result. Much more artistic control and blend happen more natural between inside and outside.

akien-mga commented on doc/classes/RenderingServer.xml:
```diff
@@ -2869,6 +2869,14 @@
 		</method>
+		<method name="reflection_probe_set_fade_start">
+			<return type="void" />
+			<argument index="0" name="probe" type="RID" />
+			<argument index="1" name="fade_start" type="float" />
```
These are now `param`.

jcostello commented:
@Calinou is this ready to merge?

Calinou commented:
> @Calinou is this ready to merge?

This needs a final review from other rendering contributors (I can't merge my own PRs).

WickedInsignia commented:
Artifacts have expired and I'd like to test this since my original proposal is being mentioned quite a lot here. I agree with ClayJohn, I'm not sure this actually resolves the proposal. Is there a build with this available somewhere?

atirut-w commented:
> Is there a build with this available somewhere?

If you dont mind running a build someone else compiled, I can make one.

WickedInsignia commented:
> > Is there a build with this available somewhere?
> 
> If you dont mind running a build someone else compiled, I can make one.
> 
> EDIT: Hold on, can the automatic builds be re-run?

When testing features I’d prefer to use an automatic or official build for security reasons. Thank you for the offer though!

Calinou commented:
I've pushed a rebase with no changes to create a new build (check back in 1 hour).

Calinou commented:
> In order to fix the underline issue, how complicated is to capture another sky reflection but with the geometry oclussion instead of using the whole sky box?

As mentioned in clayjohn's comment, you need to store a depth cubemap along the color cubemap (or perhaps store it in the cubemap's alpha channel so that no separate texture is needed). This will require slightly more time to update the cubemap but it should allow for properly occluding the sky.

This may also allow for improving parallax-corrected cubemaps with something similar to parallax occlusion mapping, as seen in God of War Ragnarok's fidelity mode.

jcostello commented:
> This will require slightly more time to update the cubemap but it should allow for properly occluding the sky.

I don't mind that requires more time. In the future, cubmaps should be baked in editor unless they are move or updated `always`

atirut-w commented:
> As mentioned in clayjohn's comment, you need to store a depth cubemap along the color cubemap (or perhaps store it in the cubemap's alpha channel so that no separate texture is needed).

Completely disabling blending when fade start is 1.0 might also work, I think

WickedInsignia commented:
In testing this definitely doesn't solve that proposal (that's not really news though). Multiple exterior probes in an interior with fade at 1 creates too many artifacts at the boundaries, and setting these probes to Interior nullifies exterior reflections.
This is a great setting to have though, and alongside probe prioritization + depth cubemaps + custom per-probe resolution would improve reflection probes considerably.

**Exterior Reflection probe, fade at 0.5:**
![ReflectionProbe_fade05](https://github.com/godotengine/godot/assets/70243134/0d02f63d-09c2-47a9-96e7-b88c6325279d)

**Exterior reflection probe, fade at 1:**
![ReflectionProbe_fade1](https://github.com/godotengine/godot/assets/70243134/1870e57e-6916-427b-ab9f-c9fd76229cd5)

**Interior reflection probe. Exterior reflections are far less present:**
![ReflectionProbe_Interior](https://github.com/godotengine/godot/assets/70243134/224f662c-92f8-4222-a51c-e43111cbc219)

**One exterior reflection probe for the bar, intersecting with the exterior reflection probe for the room. Both have a fade at 1. This artifact on the wood is too obvious and both probes would need to be interior instead:**
![ReflectionProbe_InteriorFade1](https://github.com/godotengine/godot/assets/70243134/a2f16543-c59b-44a7-bc7a-3b8c3ebc9acf)


**Couple important things I've noticed:**
- There seems to be some frustum culling of selected reflection probes in this PR. The borders and handles were disappearing at certain angles for me. This is very apparent in the Unreal Sun Temple, and doesn't happen in 4.2 Beta 2.
- This property is very difficult to edit and see its effects when the reflection probe's bounds are represented by a near-opaque green border. I might need to open a proposal to change this to something less visually obscuring.

jcostello commented:
Definitely probe prioritization and depth cubemaps should be a must to have acuarte results

atirut-w commented:
Is this PR dead? I'm a little worried considering that this (OR the depth cubemaps solution) is pretty important to have.

WickedInsignia commented:
> Is this PR dead? I'm a little worried considering that this (OR the depth cubemaps solution) is pretty important to have.

Unfortunately since the Godot project does not have individual task prioritization or a comprehensive public-facing roadmap we have no idea. I have pushed for both in the past for this very reason. It doesn't matter how in-demand a feature is: either a core team member has to be interested enough to do it or an external contributor needs to come along and implement it themselves. In the case of the latter, there's a chance it could be stuck in yet-to-be-merged limbo for a while.

I also consider this an important feature and as of right now, reflection probes aren't useful for mixed exterior/interior spaces.

Calinou commented:
Rebased and tested again, it works as expected. I've implemented the feature in Compatibility as well, following https://github.com/godotengine/godot/pull/88056.

atirut-w commented:
BOOYAH! This will probably get marked as off topic, but I'm so glad this is not dead, considering how essential of a feature this is. I hope this is getting cherry picked once its done?

Calinou commented:
> I hope this is getting cherry picked once its done?

As per the [release policy](https://docs.godotengine.org/en/latest/about/release_policy.html), new features are not cherry-picked to previous releases.

lander-vr commented:
> Please test this on more complex scenes to check if the proposed default value is actually a good idea. It may be a bit too aggressive; perhaps 0.2 would be a better fit.

I would argue for the opposite. Assuming 0.5 means half the longest length of a probe, a higher default value makes more sense. Especially with the context of the probe priority PR (#89408).

Generally you'd want whatever is within the volume of a probe to be able to fully rely on that probe by default, and just want a bit of blending towards the probe borders to avoid a sudden obvious change. Since probe workflows very commonly involve having nested probes (for which we need probe priority), a small _fade start_ would introduce a lot of duplicated reflections. A default factor of 0.5 means half of the area covered by the probe experiences those double reflections, which is essentially leakage from other probes, this is definitely something we'd want to minimize by default.

I tried to demonstrate this in a test project, it is not an ideal demonstration due to probe prioritization not being implemented yet, however it shows the effect that would happen on these transition areas between probes:
![image](https://github.com/user-attachments/assets/0aa90562-96b8-4bea-b286-775e4b04d7f7)
-> A single window turns into three reflections. (The way these probes are set up would obviously not be ideal even with probe prioritization, however it does the job of showcasing the issue)

Something like 0.75 seems like a more appropriate default value to me: You get a bit of blending at the borders as a smooth transition into the probe, and the majority of objects covered by that probe will receive clean (i.e. not double or blended) reflections.

I do think part of the reason it might feel too aggressive is because for the ambient light from probes these higher values are in fact too aggressive as a default. I think it could be worth considering having a separate fade value for the ambient light, since ambient light is a much lower frequency effect compared to reflections. This would allow great artistic control of minimizing visible seams between nested probes (especially probes placed for detailing) where you'd still want the probe to take over reflections quickly (i.e. high-fade start), but want the ambient light influence to remain subtle and blend well with the "parent" probe.

Lastly, I think it could make sense to rename and re contextualize this property. Fade start makes sense as a property name when you look at what happens under the hood, but I think from an artists POV it would be a more intuitive experience if 0 means no blending (sharp edge), and increasing this parameter increases this blending distance. The property could be name "blending" (Or reflection blending, ambient blending)

Besides these things I've brought up: the behavior of the fade start works mostly as expected, though having non-cubic probes means there are inconsistencies with blending distances on the various sides of the probe. I could imagine this be an issue for things like alleyways, corridors, etc.
Shown here using a custom ambient color on a probe:
![image](https://github.com/user-attachments/assets/a1f54262-b95d-4d72-9901-0029d941a350)
I don't think this is that big of a deal, and it shouldn't prevent this from being merged in the case that this is difficult to fix, but it is definitely something to be aware of.

ricky-daniel13 commented:
> 
> Lastly, I think it could make sense to rename and re contextualize this property. Fade start makes sense as a property name when you look at what happens under the hood, but I think from an artists POV it would be a more intuitive experience if 0 means no blending (sharp edge), and increasing this parameter increases this blending distance from the edge of the probe. The property could be name "blending" (Or reflection blending, ambient blending)
> 
I believe the concept of Fade Start is used in others places in Godot? Namely the distance fade option in lights, so maybe the name is approriate.
But yeah, I'm pretty sure the expectation of someone using reflection probes is for the probe to be the main source of reflected light on objects inside its bounds. a very tight default fade start would make sense.
(Not sure if ambient light distance should be separate, I don't imagine any reason to want different values for those)

clayjohn commented:
@lander-vr Can you clarify whether the fade start property is needed because artists need finer control over how probes blend together, or is it needed simply to avoid the nasty artifacts that we currently get as a result of blending across the entire probe?

Based on the discussion here and in https://github.com/godotengine/godot-proposals/issues/3013 it seems to me like the real problem is not that fine-tuned blending is needed, but rather we need to give artists a way to avoid having the sky and other probes blend in to the center of the current probe. Accordingly, I feel like the solution would be:
1. reduce the current blend to a much smaller margin
2. detect occlusion so we can improve the blending towards the sky.

The reason I ask is that this setting will reduce performance for every reflection probe and once we add it we can't take it back, even if we solve the root of the issue later.

lander-vr commented:
I think exposing a parameter to control blending is a must. Here's a classic situation:
![image](https://github.com/user-attachments/assets/3fb52653-fdcc-4f11-9ac7-7c816b0dab6a)
A single wall where a player can be on both sides. Considering different scenes on both sides, you want reflections of the one scene on one side, and the reflections of the other scene on the other side.

If the probe has a hard-coded blend distance, these are your only options (Blend distance portrayed by the interior boxes):
![image](https://github.com/user-attachments/assets/fd2baf25-8890-4e24-9976-3f7477b209ea)
- option A: Avoid reflection leakage on the other wall, with the drawback of not getting full reflection strength on the sides of the wall
- Option B: Overlap the probes, so the wall is far enough inside the probe to be fully affected by it. This will introduce reflection leakage.

Allowing the user to set a very narrow blend distance resolves this situation. Which is not something you'd want by default because that'd introduce seams in most situations.

Now replace that wall with some foliage:
![image](https://github.com/user-attachments/assets/b065d697-5cae-4e0a-8550-2129172123c1)
Here the user would want a much smoother blend distance, and possibly even some overlap, to avoid a visible seam running over the bushes

If the situation presented in that proposal was a situation where you'd also be able to go outside and see the other side of those walls, and you wouldn't be able to adjust the blend, you'd end up with either poor reflections, or leakage.

lander-vr commented:
> I believe the concept of Fade Start is used in others places in Godot? Namely the distance fade option in lights, so maybe the name is approriate.

I think that's a different context. The probe itself as an object is not getting faded, there's also an extra parameter there called "begin". In this case there's also already a specific distance context: the bounds of the probe. I strongly believe presenting this parameter as a setting that "blends from the edges of the probe" is a much more intuitive UX for artists.

Calinou commented:
- Superseded by https://github.com/godotengine/godot/pull/99958 (which further improves this PR).
